### PR TITLE
Support of uploading artifacts to S3 buckets

### PR DESCRIPTION
**Pull request by UnderGreen**

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -70,6 +70,15 @@ properties([
     booleanParam(name: 'BUILD_GRAVITY_HELM_APP',
                  defaultValue: false,
                  description: 'Generate a Gravity Helm App tarball'),
+    string(name: 'S3_UPLOAD_PATH',
+           defaultValue: '',
+           description: 'S3 bucket and path to upload built application image. For example "builds.example.com/cluster-ssl-app".'),
+    booleanParam(name: 'IMPORT_APP_PACKAGE',
```
Comment (federicoboerr) [inline Jenkinsfile]: Instead of `IMPORT_APP_PACKAGE` can we rename to `PUBLISH_APP_PACKAGE`?

File: Jenkinsfile [hunk 2]
```diff
@@ -154,13 +162,50 @@ node {
...
+  }
+}
+
+def isProtectedBranch(branchOrTagName) {
+  // check if tag or branch empty
+  if (!branchOrTagName?.trim()) {
+    return false
+  }
+
+  String[] protectedBranches = ['master', 'support/1.12.x']
```
Comment (federicoboerr) [inline Jenkinsfile]: the pattern for support branches can be `support/*` as all support branches are protected.

File: Jenkinsfile [hunk 3]
```diff
@@ -154,13 +162,56 @@ node {
...
+      if (isProtectedBranch(env.TAG) && params.PUBLISH_APP_PACKAGE) {
+        withCredentials([usernamePassword(credentialsId: "${AWS_CREDENTIALS}", usernameVariable: 'AWS_ACCESS_KEY_ID', passwordVariable: 'AWS_SECRET_ACCESS_KEY')]) {
+          def artifactName = "application.tar"
+          def s3AppName = "stolon-app"
+          if (params.BUILD_GRAVITY_HELM_APP) {
+            artifactName = "helm-application.tar"
+            s3AppName = "stolon-app-helm"
+          }
+          def s3Url = "s3://${S3_UPLOAD_PATH}/${s3AppName}:${APP_VERSION}.tar"
+          withEnv(MAKE_ENV + ["S3_URL=${s3Url}", "artifactName=${artifactName}"]) {
```
Comment (pablodcar) [inline Jenkinsfile]: just a detail, but is there a reason to send S3_URL and artifactName as env variables for sh to expand, you can use double quotes and replace in the command as you did for s3Url

File: Jenkinsfile [hunk 4]
```diff
@@ -154,13 +162,56 @@ node {
...
+    if (status == 0) {
+      match = true
+      return true
+    }
+  }
+
+  if (match) {
+    return true
   }
+  return false
```
Comment (pablodcar) [inline Jenkinsfile]: `any` ends early and returns `true` if it finds an element that matches the condition, so this code can be
```
return protectedBranches.any { protectedBranch ->
  if (branchOrTagName == protectedBranch) return true
  def status = sh(script: "git branch --all --contains=${branchOrTagName} | grep '[*[:space:]]*remotes/origin/${protectedBranch}\$'", returnStatus: true)
  return status == 0
}

**Review by pablodcar — CHANGES_REQUESTED** [review]

**Review by pablodcar — APPROVED** [review]

**Pull request merged**